src/internal/operators/delay.ts

```diff
@@ -59,96 +55,75 @@ import { lift } from '../util/lift';
+       * Used to check to see if we should complete the resulting
+       * subscription after delays finish or when the source completes.
+       * We don't want to complete when the source completes if we
+       * have delays in flight.
+       */
+      const checkComplete = () => isComplete && !active && !absoluteTimeValues?.length && subscriber.complete();
 
-  private _schedule(scheduler: SchedulerLike): void {
-    this.active = true;
-    const { destination } = this;
-    // TODO: The cast below seems like an issue with typings for SchedulerLike to me.
-    destination.add(
-      scheduler.schedule<DelayState<T>>(DelaySubscriber.dispatch as any, this.delay, {
-        source: this,
-        destination,
-        scheduler,
-      } as DelayState<T>)
-    );
-  }
-
-  protected _next(value: T) {
-    const scheduler = this.scheduler;
-    const message = new DelayMessage(scheduler.now() + this.delay, value);
-    this.queue.push(message);
-    if (this.active === false) {
-      this._schedule(scheduler);
-    }
-  }
-
-  protected _error(err: any) {
-    this.queue.length = 0;
-    this.destination.error(err);
-    this.unsubscribe();
-  }
-
-  protected _complete() {
-    if (this.queue.length === 0) {
-      this.destination.complete();
-    }
-    this.unsubscribe();
-  }
-}
+      if (isAbsoluteDelay) {
+        // A date was passed. We only do one delay, so let's get it
+        // scheduled right away.
+        active++;
+        subscriber.add(
+          scheduler.schedule(() => {
+            active--;
+            if (absoluteTimeValues) {
+              const values = absoluteTimeValues;
+              absoluteTimeValues = null;
+              for (const value of values) {
+                subscriber.next(value);
+              }
+            }
+            checkComplete();
+          }, +delay - scheduler.now())
+        );
+      }
 
-class DelayMessage<T> {
-  constructor(public readonly time: number, public readonly value: T) {}
+      // Subscribe to the source
+      source.subscribe(
+        new OperatorSubscriber(
+          subscriber,
+          (value) => {
+            if (isAbsoluteDelay) {
+              // If we're dealing with an absolute time (via Date) delay, then before
+              // the delay fires, the `absoluteTimeValues` array will be present, and
+              // we want to add them to that. Otherwise, if it's `null`, that is because
+              // the delay has already fired.
+              absoluteTimeValues ? absoluteTimeValues.push(value) : subscriber.next(value);
+            } else {
+              active++;
+              subscriber.add(
+                scheduler.schedule(() => {
+                  active--;
+                  subscriber.next(value);
+                  checkComplete();
+                }, delay as number)
+              );
+            }
+          },
+          undefined,
+          () => {
+            isComplete = true;
+            checkComplete();
+          }
```

Comment:
question: I guess this is another place that might warrant a teardown that nulls out the array. Thoughts?

Reply:
Good catch! Added. This PR has been really good for code health, IMO.